src/TokenHandler.mo

```diff
@@ -0,0 +1,388 @@
+    fee : Nat,
+  ) {
+
+    let icrc1Ledger = actor (Principal.toText(icrc1LedgerPrincipal)) : ICRC1.ICRC1Ledger;
+
+    // The map from p to tracking info:
+    // var tree : RBTree.RBTree<Principal, InfoLock> = RBTree.RBTree<Principal, InfoLock>(Principal.compare);
+
+    let map : Map = Map();
+    let backlog : BackLog = BackLog();
+
+    // a backlog of principals, waiting for consolidation
+    // var backlog : AssocList.AssocList<Principal, ()> = null;
+    // var backlogSize : Nat = 0;
+
+    /// query the usable balance
+    public func balance(p : Principal) : Nat = info(p).usable_balance;
+
+    /// query all tracked balances for debug purposes
+    public func info(p : Principal) : Info and {
+      var usable_balance : Nat;
+    } {
+      let ?item = map.get(p) else return {
+        var deposit = 0;
+        var credit = 0;
+        var usable_balance = 0;
+      };
+      {
+        var deposit = item.deposit;
+        var credit = item.credit;
+        var usable_balance = usableBalance(item);
+      };
+    };
+
+    /// retrieve the current size of consolidation backlog
+    public func backlogSize() : Nat = backlog.size();
+
+    /// deduct amount from P’s usable balance. Return false if the balance is insufficient.
+    public func debit(p : Principal, amount : Nat) : Bool {
+      map.change(
+        p,
+        func(info) {
+          if (usableBalance(info) < amount) return false;
+          info.credit -= amount;
+          return true;
+        },
+      );
+    };
+
+    ///  add amount to P’s usable balance (the credit is created out of thin air)
+    public func credit(p : Principal, amount : Nat) : () {
+      ignore map.set(
+        p,
+        func(info) {
+          info.credit += amount;
+          true;
+        },
+      );
+    };
+
+    /// The handler will call icrc1_balance(S:P) to query the balance. It will detect if it has increased compared
+    /// to the last balance seen. If it has increased then it will adjust the deposit (and hence the usable_balance).
+    /// It will also schedule or trigger a “consolidation”, i.e. moving the newly deposited funds from S:P to S:0.
+    public func notify(p : Principal) : async* () {
+      if (not map.lock(p)) return;
+      try {
+        let latestBalance = await* loadBalance(p);
+        updateDeposit(p, latestBalance);
+        if (latestBalance > fee) {
+          // schedule consolidation for this p
+          backlog.push(p);
+          ignore processBacklog();
+        };
+        map.unlock(p);
+      } catch err {
+        map.unlock(p);
+        throw err;
+      };
+    };
+
+    /// process first account from backlog
+    public func processBacklog() : async* () {
+      func consolidate(p : Principal) : async* () {
+        let latestBalance = try { await* loadBalance(p) } catch (err) {
+          backlog.push(p);
+          return;
+        };
+        if (latestBalance <= fee) return;
+
+        updateDeposit(p, latestBalance);
```

Comment:
updateDeposit should come before `if (latestBalance <= fee) return;` 
Even if <= fee we still want the latestBalance in the tree.